### ISSUE TITLE
Bug: Inconsistencies between nested SuspenseList and single SuspenseList

### ISSUE THREAD

**Issue by dubzzz** [facebook/react#18661]
**React version:** 0.0.0-experimental-e5d06e34b

Hoping this can be helpful for you. Here is what looks to be a bug with concurrent mode and nested SuspenseList.

## Steps To Reproduce

_In concurrent mode_

0. Setup

Let's suppose we have three components `<A />` (**not** lazy loaded), `<B />` (**not** lazy loaded) and `<C />` (lazy loaded).

In other words:

```jsx
import A from './A';
import B from './B';
const C = React.lazy(() => import('./C'));
```

1. Render

```jsx
render(
  <SuspenseList key="1" revealOrder="forwards">
    <SuspenseList key="1.1" revealOrder="forwards">
      <Suspense key="1.1.a" fallback={<Text text="Loading A" />}>
        <A />
      </Suspense>
    </SuspenseList>
  </SuspenseList>
)
```

2. Update the component (component now shows A, B, and C)

```jsx
render(
  <SuspenseList key="1" revealOrder="forwards">
    <SuspenseList key="1.1" revealOrder="forwards">
      <Suspense key="1.1.a" fallback={<Text text="Loading A" />}>
        <A />
      </Suspense>
      <Suspense key="1.1.b" fallback={<Text text="Loading B" />}>
        <B />
      </Suspense>
      <Suspense key="1.1.c" fallback={<Text text="Loading C" />}>
        <C />
      </Suspense>
    </SuspenseList>
  </SuspenseList>
)
```

3. Output is: `A / Loading B / Loading C`. While `B` has already been loaded (not lazy loaded). If I understand well the behaviour of `forwards` I would have expect to have `A / B / Loading C` instead.

Please note that the behaviour is not the same if I do not use nested `<SuspenseList />`.

Live repro: https://dubzzz.github.io/react-suspenselist-bug/build/
Link to code example (started with CRA): https://github.com/dubzzz/react-suspenselist-bug

## The current behavior

Output is: `A / Loading B / Loading C` with nested `SuspenseList`
Output is: `A / B / Loading C` without nested `SuspenseList`

## The expected behavior

Output is: `A / B / Loading C` with both

**How did I found this bug?**

This _potential bug_ has been discovered while I was trying to run property based tests based on [fast-check](https://github.com/dubzzz/fast-check/) against React library.

See https://github.com/dubzzz/react/commit/e2cb4776ea3634fab2021d83cb8259bde03c0f3a

**Comment by gaearon** [comment]
Haven’t looked at the issue yet (and might not for a bit). But we need a lot stronger fuzz coverage for Suspense, SuspenseList, and priority scheduling. See the recent closed bugs with “concurrent mode” label for examples of how things go wrong and what catches them. If you’d like to contribute, coming up with a stronger fuzz and submitting failing tests would be much appreciated.

**Comment by gaearon** [comment]
For SuspenseList specifically I don’t remember what the semantics for *updating* are. There was something subtle there. So maybe this isn’t a bug per se.

**Comment by gaearon** [comment]
I agree with you this looks weird though. Can you submit an isolated failing case PR for now that’s manually written?

And then add fuzz separately.

**Comment by gaearon** [comment]
I also opened https://github.com/facebook/react/issues/18669 for the broader fuzz question.

**Comment by dubzzz** [comment]
> Can you submit an isolated failing case PR for now that’s manually written?

Yes, I will open a specific PR only containing a failing unit test related to this potential bug.

> And then add fuzz separately.

And then open PRs to discuss fuzzing on SuspenseList. And link them to #18669.

**Comment by sebmarkbage** [comment]
Updates are indeed strange. Nesting is also strange.

What happens here is that we have two rules that can yield counter intuitive results:

- A row is not allowed to show content until the previous row is done.
- A row is not allowed to show partial content until the whole row is done.
- Never hide already shown content until it explicitly resuspends, to avoid content disappearing.

What's happening here is that the inner Suspense list is also a row. What happens here is the rule: "A row is not allowed to show partial content until the whole row is done."

It gets extra confusing because A is still showing due to the rule "Never hide already shown content until it explicitly resuspends, to avoid content disappearing".

So this behavior is "intentional" according to the rules.

We might have the wrong heuristics but a real world use case would be helpful to inform how they should change.

When you have two Suspense boundaries in a row that aren't wrapped in a SuspenseList there has to be kind of an implicit list there to decide how they should work with regard to each other. You wouldn't want the last one to pop in before the first one. So a good default is "A row is not allowed to show partial content until the whole row is done". However, I could imagine a rule like if all boundaries are in a single SuspenseList in a row, then that list decides.

**Comment by dubzzz** [comment]
@sebmarkbage 
Thanks a lot for the explanation. Now I think I understand part of the behaviour I see:
- We see A in the second render because of the rule: _Never hide already shown content until it explicitly resuspends, to avoid content disappearing_

There is still something I don't fully get - Why is the behaviour different if I directly render:
```js
render(
  <SuspenseList key="1" revealOrder="forwards">
    <SuspenseList key="1.1" revealOrder="forwards">
      <Suspense key="1.1.a" fallback={<div>Loading A</div>}>
        <A />
      </Suspense>
      <Suspense key="1.1.b" fallback={<div>Loading B</div>}>
        <B />
      </Suspense>
      <Suspense key="1.1.c" fallback={<div>Loading C</div>}>
        <C />
      </Suspense>
    </SuspenseList>
  </SuspenseList>
)
```
Case of the issue: https://codesandbox.io/s/mutable-rain-3ikor?file=/src/App.js (click on update)
Case with direct render: https://codesandbox.io/s/trusting-hoover-4cffd?file=/src/App.js

I mean according to your explanation I'd have expected to see `Loading A / Loading B / Loading C` but I see `A / B / Loading C` instead 🤔 

(What is a `row`? A `Suspense`? A `SuspenseList`? Both? Something else?)

**Comment by gaearon** [comment]
I think a "row" is an immediate child of SuspenseList?

**Comment by sebmarkbage** [comment]
@dubzzz In your example I see `A / Loading B / Loading C`.

That's because we are not allowed to turn A back into "Loading A" because:
_"Never hide already shown content until it explicitly resuspends, to avoid content disappearing."_

We're not allowed to show "B" because:
_"A row is not allowed to show partial content until the whole row is done."_.
The row in this case is the whole list because it's a row of the outer list.

**Comment by dubzzz** [comment]
@sebmarkbage Thanks a lot for this very detailed explanation. I get a better idea of how updates are supposed to be working.

**Maybe one last question: Do these rules only apply on updates? (not first render)**

Indeed, they seem not to apply on first render. See the following example not behaving the same way?

> https://codesandbox.io/s/trusting-hoover-4cffd?file=/src/App.js



In this second example I directly render: (with A, B already resolved BUT C still pending)
```jsx
render(
  <SuspenseList key="1" revealOrder="forwards">
    <SuspenseList key="1.1" revealOrder="forwards">
      <Suspense key="1.1.a" fallback={<div>Loading A</div>}>
        <A />
      </Suspense>
      <Suspense key="1.1.b" fallback={<div>Loading B</div>}>
        <B />
      </Suspense>
      <Suspense key="1.1.c" fallback={<div>Loading C</div>}>
        <C />
      </Suspense>
    </SuspenseList>
  </SuspenseList>
)
```

So given your explanation I expect to have `Loading A / Loading B / Loading C`.

Indeed neither A, nor B have already been displayed so "_Never hide already shown content until it explicitly resuspends, to avoid content disappearing._" does not apply (first render).

Additionally as "_A row is not allowed to show partial content until the whole row is done._", nothing from the row should be shown as C is still loading.

As a consequence of those two rules I expect to have `Loading A / Loading B / Loading C` but I get `A / B / Loading C`.

**Comment by sebmarkbage** [comment]
Yea that one is kind of surprising to me too.

Regardless I came across a case where we might want to change the rules like I mentioned above.

If you have a row that you want to consistently show in its loading state while popping in additional items one by one before it, then you need an inner list with tail=collapsed/hidden and an outer one which is not collapsed. But we force the last item not to show its fallback.

However if you do that then the inner one isn’t allowed to partially suspend.

So regardless I think we need to change this.

**Comment by hosseinmd** [comment]
What is the latest version of concurrent mode?
